### ISSUE TITLE
[Feature Request] Use read-write lock instead of global lock when sampling is non-mutating

### ISSUE THREAD

**Issue by adityagoel4512** [pytorch/rl#604]
## Motivation

The use of threading.RLock() in ReplayBuffer prototypes limits the number of concurrent reads that can be performed on a replay buffer unnecessarily when contents are in shared memory. Using a global reentrant lock for synchronising access to the replay buffer can cause more contention than necessary. Concurrent reads are permissable for many replay buffer/storage types where sampling does not mutate the underlying buffer, and can improve performance in distributed RL tasks. 

## Solution

Switching to using a read-write lock resolves this by allowing multiple concurrent reads (buffer samples), making sampling blocking only if the write lock (for inserting into the replay buffer) is acquired.

## Alternatives

We could introduce a constructor parameter to allow advanced users to specify an appropriate synchronisation strategy. 

## Additional context

https://github.com/pytorch/rl/blob/main/torchrl/data/replay_buffers/rb_prototype.py
https://github.com/pytorch/rl/blob/main/torchrl/data/replay_buffers/replay_buffers.py

Not all replay buffers should use read-write locks - this applies only for when sampling is a non-mutating function.

## Checklist

- [ ] I have checked that there is no similar issue in the repo (**required**)

**Comment by vmoens** [comment]
This was more than a year ago but we should fix this!
@xiaomengy do you have a suggestion? Would you like to help with this?

**Comment by xiaomengy** [comment]
Yes, I'm happy to help on this. But I'm very busy this week. May I take a look in next week?
